akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/ConsumerFilterRegistry.scala

```diff
@@ -35,27 +50,113 @@ import akka.projection.grpc.consumer.ConsumerFilter
-        //       streamId. Is it important to be able to filter those separately?
+  private def behavior(
+      subscribers: Map[Subscriber, immutable.Seq[FilterCriteria]],
+      stores: Map[String, ActorRef[ConsumerFilterStore.Command]]): Behavior[Command] = {
 
-        behavior(subscribers + Subscriber(sub.streamId, sub.subscriber))
+    def getOrCreateStore(streamId: String): ActorRef[ConsumerFilterStore.Command] = {
+      stores.get(streamId) match {
+        case Some(store) => store
+        case None =>
+          context.spawn(
+            ConsumerFilterStore(context.system, streamId, context.self),
+            URLEncoder.encode(streamId, StandardCharsets.UTF_8.name))
+      }
+    }
 
-      case cmd: SubscriberCommand =>
-        // FIXME the registered filters (aggregated) must also be kept in state and distributed to other nodes in the cluster
+    def publishUpdatedFilterToSubscribers(
+        streamId: String,
+        filter: immutable.Seq[FilterCriteria]): Map[Subscriber, immutable.Seq[FilterCriteria]] = {
+      subscribers.map {
+        case (sub, subFilter) =>
+          if (sub.streamId == streamId) {
+            val diff = ConsumerFilter.createDiff(subFilter, filter)
+            if (diff.nonEmpty)
+              sub.ref ! UpdateFilter(streamId, diff)
+            sub -> filter
+          } else {
+            sub -> subFilter
+          }
+      }
+    }
 
-        context.log.debug("Publish command [{}] to subscribers", cmd)
-        subscribers.foreach { sub =>
+    def publishToSubscribers(cmd: SubscriberCommand): Unit = {
+      context.log.debug("Publish command [{}] to subscribers", cmd)
+      subscribers.foreach {
+        case (sub, _) =>
           if (sub.streamId == cmd.streamId)
             sub.ref ! cmd
-        }
-        Behaviors.same
+      }
     }
+
+    Behaviors
+      .receiveMessage {
+        case Subscribe(streamId, initCriteria, subscriberRef) =>
+          val subscriber = Subscriber(streamId, subscriberRef)
+          context.watchWith(subscriberRef, SubscriberTerminated(subscriber))
+          // eagerly start the store
+          val store = getOrCreateStore(streamId)
+
+          // trigger a refresh from current filter so that updates between the initCriteria and the Subscribe is flushed
+          // to the new subscriber
+          implicit val askTimeout: Timeout = 10.seconds
+          context
+            .ask[ConsumerFilterStore.GetFilter, ConsumerFilter.CurrentFilter](store, ConsumerFilterStore.GetFilter(_)) {
+              case Success(ConsumerFilter.CurrentFilter(_, criteria)) => FilterUpdated(streamId, criteria)
+              case Failure(_) =>
+                context.log
+                  .debug("Ask of current filter failed when subscriber for streamId [{}] registered.", streamId)
+                null // ignore, it will be ok on next update anyway
```

Comment:
FYI, not likely to happen but, returning null here will log a warning in the actor adapter adapt message handling

Reply:
um, could add another message, but this local ask shouldn't really fail

Reply:
Yeah, let's leave it as is (it just made me wonder so I looked up what will happen)